### DISCUSSION TITLE
PicoCTF Cache Me Outside Write Up

### DISCUSSION THREAD

**Discussion by hehirm** [Dvd848/CTFs#29] [General]
Hi,

Firstly I would like to say that your write up for the picoCTF Cache Me Outside has been instrumental in my journey into heap exploitation. Your solution is the only one I have been able to replicate and fully understand.

In the final stages of the exploit you pass the following command to the program:

`{ echo "-5144"; printf "\x00";}`

This produces the desired result. However, I have found that if you instead simply run the program and enter the results manually (instead of using a pipe) the program doesn't spit out the flag as intended. It seems as though doing it this way changes the tcache pointer value to be null, which doesn't produce any output in the final call to puts.

I would greatly appreciate some more guidance on why the above command was chosen, considering the other ways shortcomings.

Thanks,

Maxim

**Comment by Dvd848** [comment]
Hi Maxim,

Thank you so much for your feedback, it's really great to hear that the writeups are helpful.

It's been a while since I dived into this challenge, but the first thing that stood out to me is the fact that `printf "\x00"` prints the actual NUL byte, which while possible, is not straightforward to do manually. When entering the input manually, are use entering the NUL byte (0x00 in ASCII) or the character "0" (0x30 in ASCII)?

**Reply by hehirm** [reply]
Hey David,

This was precisely the issue - after a bit of google searching I was able to find that `Ctrl` + `v` followed by `Ctrl` + `Shift` + `@` was able to send the input through the program. However it is nice to know there is a way of doing it that doesn't involve memorizing key bindings.

Thank you for your reply and all the best in your programming endeavours
